highfidelity/hifi
# Replace vec2, vec3, and xColor properties with JS prototypes

Fixes:
https://highfidelity.manuscript.com/f/cases/16341/vec3-saving-with-blue-green-red-info-in-json
https://highfidelity.manuscript.com/f/cases/10596/Entities-editEntity-color-changes-should-support-r-g-b-in-addition-to-spelling-out-red-green-blue
https://highfidelity.manuscript.com/f/cases/12274/PolyLine-entity-s-strokeColors-property-should-be-an-array-of-color-values-not-Vec3-values
https://highfidelity.manuscript.com/f/cases/16358/Entity-color-edits-not-sticking
https://highfidelity.manuscript.com/f/cases/17073/MyAvatar-position-returns-an-object-with-keys-red-green-blue-and-x-y-z
https://highfidelity.manuscript.com/f/cases/17003/Export-Entity-JSON-file-exports-both-xyz-AND-rgb

Before, our JS vec3s were just maps with x, y, and z properties.  There was also a color type with only red, green, and blue.  Now, vec2s and vec3s are exposed to JS as objects with properties (x, y, z) and aliases (r, g, b)/(red, green, blue).  Colors will stringify with red/green/blue and other vectors will stringify with x/y/z.  Similarly, vec2s have (x, y) and alias (u, v).

What does this mean?

Before:
```
var color = Entities.getEntityProperties(<entity id>).color;
print(JSON.stringify(color)); // prints `{red: 255, green: 0, blue: 0}`
color.r = 127;
print(JSON.stringify(color)); // prints `{red: 255, green: 0, blue: 0, r: 127}`
Entities.editEntity(<entity id>, { color: {red: 0, green: 255, blue: 0} }); // works
Entities.editEntity(<entity id>, { color: {r: 0, g: 255, b: 0} }); // doesn't work
Entities.editEntity(<entity id>, { color: {x: 0, y: 255, z: 0} }); // doesn't work
Entities.editEntity(<entity id>, { color: [0, 255, 0]); // doesn't work
```
Now:
```
var color = Entities.getEntityProperties(<entity id>).color;
print(JSON.stringify(color)); // prints `{red: 255, green: 0, blue: 0}`
color.r = 127;
print(JSON.stringify(color)); // prints `{red: 127, green: 0, blue: 0}`
Entities.editEntity(<entity id>, { color: {red: 0, green: 255, blue: 0} }); // works
Entities.editEntity(<entity id>, { color: {r: 0, g: 255, b: 0} }); // works
Entities.editEntity(<entity id>, { color: {x: 0, y: 255, z: 0} }); // works
Entities.editEntity(<entity id>, { color: [0, 255, 0]); // works
```

Test plan:
- These changes affect a lot of different areas.  Run the autotester to confirm that the tests in hifi_tests still work.
- Verify that Create still works properly, with special attention to vec2 or vec3 properties: position, dimensions, various physics properties, and particle properties (verify that the particle tab works correctly, especially the color fields and color pickers)
- Export some entities (try: primitives and particles after messing with their properties).  In the resulting JSON, confirm that vec3 properties only list x, y, and z (not both xyz and red/green/blue).  Colors should list red/green/blue (check: cube color, particle color/colorSpread/colorStart/colorFinish, text textColor/backgroundColor, zone skybox color/haze color/keylight color).  Import those entities again and confirm that the properties persisted correctly.
- Confirm that you can edit an entity's color like this: `Entities.editEntity(<entity id>, { color: { x: 255, green: 155, b: 55 } });`.  Go into create and confirm that its color is `{ red: 255, green: 155, blue: 55 }`
- Create a sphere.  Sphere is red. Shut down Interface and server.  Restart server and Interface.  Edit sphere's color: change green component to 255, then change blue component to 255.  Sphere is white. 
 Restart server and Interface.  Sphere should be white (not red, as in master).
- In the console, enter `JSON.stringify(MyAvatar.position)`.  The result should only have xyz.
- Verify that Polylines still work.  Try the Fingerpaint app.
